lib/web/saml.go

```diff
@@ -97,8 +99,22 @@ func (h *Handler) samlACS(w http.ResponseWriter, r *http.Request, p httprouter.P
 	}
 
 	response, err := h.cfg.ProxyClient.ValidateSAMLResponse(samlResponse)
+
 	if err != nil {
 		logger.WithError(err).Error("Error while processing callback.")
+
+		// try to find the auth request, which bears the original client redirect URL.
+		// if found, use it to terminate the flow.
+		//
+		// this improves the UX by terminating the failed SSO flow immediately, rather than hoping for a timeout.
```

Comment:
Hm, I'm not sure why we the valid redirect URL is returned here instead of old `LoginFailedBadCallbackRedirectURL` url and where the actually timeout occurs. 
If the `LoginFailedBadCallbackRedirectURL` is returned the client should be able call local url instantly and exit failed SSO flow but if the request url.String() can be malformed and not reachable bu the client.  If some additional information needs to be passed to the client why not just  encode them in the same way  in `LoginFailedBadCallbackRedirectURL` ?

Reply:
If the SSO login is initiated from the console (e.g. `tsh login`) then ending the flow on `/web/msg/error/login/callback` instead of `ClientRedirectURL` causes the console to leave it hanging.

Eventually 3 minutes pass and [this happens](https://github.com/gravitational/teleport/blob/master/lib/client/weblogin.go#L319):

```go
	case <-time.After(defaults.CallbackTimeout):
		log.Debugf("Timed out waiting for callback after %v.", defaults.CallbackTimeout)
		return nil, trace.Wrap(trace.Errorf("timed out waiting for callback"))
```

The browser will get redirected to error URL by `Redirector` in `wrapCallback`: https://github.com/gravitational/teleport/blob/master/lib/client/redirect.go#L231

I'm not sure what is the exact change that you are proposing here?

Reply:
Good catch! Fixed: https://github.com/gravitational/teleport/pull/11508/commits/2ed6ad434a8263c53dff3bf60eae5c9ed10fd7b3